x/stake/keeper.go

```diff
@@ -167,48 +172,60 @@ func (k Keeper) GetValidators(ctx sdk.Context) (validators []Validator) {
 	iterator := store.Iterator(subspace(RecentValidatorsKey))
 	for ; iterator.Valid(); iterator.Next() {
 		addr := AddrFromKey(iterator.Key())
-		store.Set(GetToKickOutValidatorKey(addr), []byte{})
+
+		// iterator.Value is the validator object
+		store.Set(GetToKickOutValidatorKey(addr), iterator.Value())
 		store.Delete(iterator.Key())
 	}
 	iterator.Close()
 
 	// add the actual validator power sorted store
-	maxVal := k.GetParams(ctx).MaxValidators
+	maxValidators := k.GetParams(ctx).MaxValidators
 	iterator = store.ReverseIterator(subspace(ValidatorsKey)) // largest to smallest
-	validators = make([]Validator, maxVal)
+	validators = make([]Validator, maxValidators)
 	i := 0
 	for ; ; i++ {
-		if !iterator.Valid() || i > int(maxVal-1) {
+		if !iterator.Valid() || i > int(maxValidators-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
-		var val Validator
-		err := k.cdc.UnmarshalBinary(bz, &val)
+		var validator Validator
+		err := k.cdc.UnmarshalBinary(bz, &validator)
 		if err != nil {
 			panic(err)
 		}
-		validators[i] = val
+		validators[i] = validator
```

Comment:
This function claims to return most recent updated validator set, but this loop iterates over the validators in order of power descending, then the function returns the head of that array?

I presume you mean the current `validator set`, as opposed to the set of the most recently updated validators (which would be retrieved using `GetRecentValidatorsKey`?) - maybe worth clarifying comment?

Reply:
Good point, I will expand the comments - the most recent updated validator set IS the "top" candidates... there are no other validators - the "Recent Validators" is just a duplicate of the results from the last time `GetValidators` was called - this way we can easily determine if a candidate is a validator without needing to iterate over the subspace as we do in GetValidators